core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
+        MAX_HEARTBEAT_MISS_ALLOWED times continuously */
+        /* Global reference count is incremented whenever heartbeat is attempted */
+        private int refCount;
+        private int refCountOnHeartBeat; /* Reference count upon successful heartbeats */
+
+        /* Minimum samples to consider data rates as consistent. */
+        private static final int MIN_STABLE_DATA_RATE_TIMES = 10;
+        private int stableDataRateTimes; /* Number of continuous stable data transfer rates */
+        private RandomData data; /* Amount of random data sent to server in heartbeats */
+
+        private int metricPollPeriod = MIN_METRIC_POLL_PERIOD_MS; /* Poll period */
+        private NodeMetric metric; /* Node metric to server */
+        private ResettableTimer metricsTimer; /* Metric timer for the server */
+
+        private KernelServerInfo(
+                InetSocketAddress serverAddress, KernelServer remoteRef, final NodeMetric metric) {
+            this.serverAddress = serverAddress;
+            this.remoteRef = remoteRef;
+            this.metric = metric;
+            data = new RandomData();
+            metricsTimer =
+                    new ResettableTimer(
+                            new TimerTask() {
+                                public void run() {
+                                    measureMetrics(KernelServerInfo.this);
+                                }
+                            },
+                            metricPollPeriod);
+            metricsTimer.start();
+        }
+    }
+
+    /** Map of remote kernel server hostname to their kernel server info */
+    private ConcurrentHashMap<InetSocketAddress, KernelServerInfo> servers;
+
+    /* Map of remote kernel server hostname to their metrics. It is maintained independent from servers map, so that
+    this map can be serialized and sent to OMS as is */
+    private ConcurrentHashMap<InetSocketAddress, NodeMetric> metrics;
 
     private static final Logger logger = Logger.getLogger(KernelClient.class.getName());
 
+    public ConcurrentHashMap<InetSocketAddress, NodeMetric> getMetrics() {
+        return metrics;
+    }
+
     /**
-     * Add a host to the list of hosts that we've contacted
+     * Adds remote kernel server information to cache for the given host
      *
      * @param host
      */
-    private KernelServer addHost(InetSocketAddress host) {
+    private KernelServerInfo addHost(InetSocketAddress host) {
         try {
             Registry registry = LocateRegistry.getRegistry(host.getHostName(), host.getPort());
             KernelServer server = (KernelServer) registry.lookup("io.amino.run.kernelserver");
-            servers.put(host, server);
-            return server;
+            NodeMetric metric = new NodeMetric();
+            KernelServerInfo serverInfo = new KernelServerInfo(host, server, metric);
+            servers.put(host, serverInfo);
+            metrics.put(host, metric);
```

Comment:
See above comment.  Merging would remove this duplication.

Reply:
Modified